### PR TITLE
Remove the newly generated key if sc add-gcp-broker command fails

### LINKED ISSUE #38: Need to clean up service account json key in "remove-gcp-broker"
In the "add-gcp-broker" sub-command, a service account is created (if it doesn't already exist) and a JSON key for this service account is created and downloaded. Currently, we don't remove/delete this key from the service account when we call "remove-gcp-broker". Only about a dozen keys are allowed for a service account, so we eventually run out of keys for the service account and throw an error.

### PR DESCRIPTION
**Pull request by maqiuyujoyce**
Fixes #38

### CODE REVIEW

File: installer/pkg/cmd/gcp_broker.go [hunk 1]
```diff
@@ -126,11 +129,15 @@ func addGCPBroker() error {
...
+		cleanupNewKey(brokerSAEmail, key)
 		return fmt.Errorf("error generating configs for GCP :: %v", err)
 	}
 
 	err = deployGCPBrokerConfigs(dir)
 	if err != nil {
+		// Clean up the newly generated key if the command failed.
+		cleanupNewKey(brokerSAEmail, key)
 		return fmt.Errorf("error deploying GCP broker configs :%v", err)
 	}
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply installer/pkg/cmd/gcp_broker.go]: Do you mean to group getOrCreateVirtualBroker(), generateGCPBrokerConfigs(), and deployGCPBrokerConfigs() together?

In that case, I'd prefer to return the key to the NewXXXCommand method and let it to handle the cleanup when the key is not empty.

File: installer/pkg/cmd/gcp_broker.go [hunk 2]
```diff
@@ -280,6 +287,25 @@ func httpAdapterFromAuthKey(keyFile string) (*adapter.HttpAdapter, error) {
 	return adapter.NewHttpAdapter(client), nil
 }
 
+// cleanupNewKey removes the newly generated service account key.
+func cleanupNewKey(email, key string) {
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply installer/pkg/cmd/gcp_broker.go]: This one is more than cleanup the "given key". It is to cleanup the given keyfile content in base64 encoding. I'll update the comment to be more specific.

File: installer/pkg/cmd/gcp_broker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mihnjong) [reply installer/pkg/cmd/gcp_broker.go]: Calling cleanupNewKey in multiple places makes further change prone to introduce a bug. Some simple option of avoiding it is creating one helper function where you put all the cases that should call cleanupNewKey(). If you have other way to achieve it, please feel free to try that out.

[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply installer/pkg/cmd/gcp_broker.go]: Will keep as is and clean up/refactor the entire tool when we have time. Tracked in #149.

File: installer/pkg/cmd/gcp_broker.go [hunk 3]
```diff
@@ -280,6 +287,25 @@ func httpAdapterFromAuthKey(keyFile string) (*adapter.HttpAdapter, error) {
...
+	}
+
+	keyJson := make(map[string]interface{})
+	err = json.Unmarshal(keyBytes, &keyJson)
+	if err != nil {
+		// Silently return if there is an error unmarshalling the key.
+		return
+	}
+
+	keyID := keyJson["private_key_id"].(string)
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply installer/pkg/cmd/gcp_broker.go]: Tracked in #149.

File: installer/pkg/gcp/broker.go [hunk 4]
```diff
@@ -182,18 +184,23 @@ func RemoveServiceAccountKeys(email string) error {
...
 	return nil
 }
 
+// RemoveServiceAccountKey removes the given key from the service account.
+func RemoveServiceAccountKey(email, keyID string) {
+	cmd := exec.Command("gcloud", "iam", "service-accounts", "keys", "delete", keyID, "--iam-account", email, "--quiet" /*disable interactive mode*/)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Printf("failed to delete service account key: %s : %v\n", string(out), err)
+		fmt.Printf("WARNING: Please clean up the key from service account %s", email)
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply installer/pkg/gcp/broker.go]: The current error message is actually "freeform" and based on the experience of each developer. We need to do a bit design and clean it up. Tracked by #150.

**Pull request merged**